Extension/browser/safari/lib/content-script/content-script.js

```diff
@@ -16,144 +16,216 @@
+			};
 
-	var contentScriptId = Date.now() + Math.random().toString(10).slice(2);
-
-	var absoluteUrlHelper = document.createElement("a");
-
-	var onFirstLoadOccurred = false;
-
-	var execTmpScript = function () {
-		var tmpJS = document.createElement("script");
-		tmpJS.textContent = '(function () {\
-								var block = function (url, type) {\
-									var event = new CustomEvent("' + contentScriptId + '", {\
-										detail: {\
-											url: url,\
-											type: type\
-										},\
-										bubbles: false\
-									});\
-									document.dispatchEvent(event);\
-									return event.detail.url === false;\
-								};\
-								var _emptyFunc = function () {\
-								};\
-								var xmlHttpRequestOpen = XMLHttpRequest.prototype.open;\
-								XMLHttpRequest.prototype.open = function (method, url) {\
-									if (block(url, "xmlhttprequest")) {\
-										return {send: _emptyFunc}\
-									} else {\
-										return xmlHttpRequestOpen.apply(this, arguments);\
-									}\
-								}\
-							})();';
-		document.documentElement.removeChild(document.documentElement.appendChild(tmpJS));
-	};
+			var left = new Version(leftVersion);
+			var right = new Version(rightVersion);
+			return left.compare(right) >= 0;
+		};
 
-	var canLoadRequest = function (url, type, frameId) {
-		return safari.self.tab.canLoad(event, {
-			type: "safariWebRequest", data: {
-				url: url,
-				type: type,
-				frameId: frameId,
-				requestFrameId: 0
-			}
-		});
+		var version = parseSafariVersion();
+		return isGreaterOrEqualsVersion(version, "9.0");
 	};
+	var safariNew = isSafariNew();
+
+	if (window.top === window) {
 
-	var onBeforeLoad = function (event) {
+        if (safariNew) {
 
-		if (!onFirstLoadOccurred) {
-			onFirstLoad();
-		}
+			function createMainFrameEvent(type) {
+				var data = {
+					url: document.location.href,
+					type: "main_frame",
+					frameId: 0
+				};
+
+				safari.self.tab.dispatchMessage('canLoad', {
+					type: type,
+					data: data
+				});
+			};
+
+			createMainFrameEvent("safariWebRequest");
+			createMainFrameEvent("safariHeadersRequest");
+
+		} else {
 
-		absoluteUrlHelper.href = event.url;
-		var url = absoluteUrlHelper.href;
+			safari.self.tab.dispatchMessage("loading", document.location.href);
 
-		if (!/^https?:/.test(url)) {
-			return;
+			function createMainFrameEvent(type) {
+				var data = {
+					url: document.location.href,
+					type: "main_frame",
+					frameId: 0
+				};
+				var evt = document.createEvent("Event");
+				evt.initEvent("beforeload");
+				safari.self.tab.canLoad(evt, {type: type, data: data});
+			}
+
+			createMainFrameEvent("safariWebRequest");
+			createMainFrameEvent("safariHeadersRequest");
 		}
+	}
 
-		var type;
-		switch (event.target.localName) {
-			case "link":
-				if (/(^|\s)stylesheet($|\s)/i.test(event.target.rel)) {
-					type = "stylesheet";
-					break;
+	if (!safariNew) {
+
+		var contentScriptId = Date.now() + Math.random().toString(10).slice(2);
```

Comment:
It would be MUCH better if you separate old and new functions into two different files.

```
/content-script-legacy.js
/content-script-content-blocker.js
```

Add both of these files to content scripts list and use functions from these files in common "content-script.js"